File: 2681-put-marbles-in-bags/2681-put-marbles-in-bags.java
class Solution {
    public long putMarbles(int[] weights, int k) {
        int n = weights.length;
        var p = new int[n - 1];
        for (int i = 0; i < n - 1; ++i)
            p[i] += weights[i] + weights[i + 1];
        
        Arrays.sort(p);
        long ans = 0;

        for (int i = 0; i < k - 1; ++i)
            ans += p[n - i - 2] - p[i];
        return ans;
    }
}

File: 2681-put-marbles-in-bags/2681-put-marbles-in-bags.cpp
using ll = long long;
using head = int;
#define v.begin(), v.end() all(v)

class Solution {
public:
    ll putMarbles(vector<int>& weights, int k) {
        head n = weights.size();
        vector<head> p(n - 1, 0);
        for (head i = 0; i < n - 1; ++i)
            p[i] += weights[i] + weights[i + 1];

        sort(p.begin(), p.end());

        ll ans = 0;
        for (head i = 0; i < k - 1; ++i)
            ans += p[n - 2 - i] - p[i];

        return ans;
    }
};